### PR TITLE
Be socket io

### PR DESCRIPTION
**Pull request by ChianHuei**
1.The BE io listens on "connection".
When user logging in, the client end creates a socket. After this socket connects to BE and the user is authenticated, client end sends a message with user id to BE. BE put this socket into a "Room" tagged with user id. We can target our message on a specific user via the Room. Furthermore, if the user opens multiple browser tabs, there will be multiple sockets in the Room.   All the tabs are able to receive the same message from BE.
When we want to send a message from BE, we may call io like this
```
req.app.io.to(foundList.user.id).emit('price_notification', { 
	message: foundList.products[0],
}); 
```

2.FE UI
When messages come in, the badge shows the number of new messages. After user opening the message box, these FE message box will be cleaned out.
Screen Shot 2020-07-15 at 8 52 26 PM

### CODE REVIEW

File: client/src/components/Navbar.js [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 import { Link as RouterLink } from 'react-router-dom';
 import LocalMallIcon from '@material-ui/icons/LocalMall';
 import AuthContext from '../state_management/AuthContext';
 import {fetchShLists} from '../state_management/actionCreators/shoppingListsActs';
 import ShListsContext from '../state_management/ShListsContext';
+import socketIOClient from "socket.io-client";
+import Notifications from "./Notifications";
+
+
+const ENDPOINT = "http://localhost:3001";
```
Comment (surhud004) [inline client/src/components/Navbar.js]: This ENDPOINT will change on deploy in prod. We will need to add this to .env

**Review by surhud004 — APPROVED** [review]
Looks good and clean!

File: client/src/components/Notifications.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            <List className={classes.list}>
+                {props.messages.map((product) => (
+                    <div key={product._id} className={classes.listItem}>
+                        <Typography color="textPrimary" variant="body1">
+                            New price!
+                        </Typography>
+                        <ListItem >
+                            <ListItemAvatar>
+                            <Avatar alt="no img" src={product.image} variant="square" type="image">
+                                P
```
Comment (sina-jamshidi) [inline client/src/components/Notifications.js]: This is nice. I wonder if just for a little bit more detail we can use the first letter of the product name if it's not null

File: client/src/components/Notifications.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                <Typography className={classes.title} color="textPrimary" variant="body2">
+                                    {cutContentLength(product.name,30,"no product name")}
+                                </Typography>
+                                <Typography component="a" href={product.url} target="_blank" rel="noreferrer" variant="caption" color="textSecondary">
+                                    {cutContentLength(product.url,30,"no link")}
+                                </Typography >
+                                <Grid container alignItems='baseline'>
+                                    <Grid item xs={6}>
+                                        <Typography color="textSecondary" style={linethrough} variant="body2">
+                                            ${covertNumberDecimal(product.lastprice)}{" "}
```
Comment (sina-jamshidi) [inline client/src/components/Notifications.js]: there's a typo in this function name `convert`

**Review by sina-jamshidi — COMMENTED** [review]
Looks good! Just a couple of notes

**Comment by sina-jamshidi** [comment]
Actually a couple more notes lol:

- Can you try making the product name bold and see how that looks? In the mocks it is bolded and I think it looks a bit clearer.
- Can you make it so that if you click the product div the product url opens in a new tab? (so the amazon link)

File: client/src/components/Navbar.js (hunk 1 above)
Comment (Ethorf) [reply client/src/components/Navbar.js]: Good catch!

File: client/src/components/Notifications.js (hunk 2 above)
Comment (ChianHuei) [reply client/src/components/Notifications.js]: I will change it.

File: client/src/components/Notifications.js (hunk 3 above)
Comment (ChianHuei) [reply client/src/components/Notifications.js]: Yeah!ha ha, I will correct it.

File: client/src/components/Navbar.js [hunk 4]
```diff
@@ -26,25 +31,73 @@ const useStyles = makeStyles((theme) => ({
...
+                setNewMsg(data.message);
+                msgHasBeenRead = false;
+            });
+            socket.emit('join_room', {
+                message: "",
+                userId: authContext._id,
+            });
+            setSocket({socket:socket});
+        }
+        if(newMsg && newMsg !== null){
```
Comment (Ethorf) [inline client/src/components/Navbar.js]: is it redundant here to check for both conditions?

**Review by Ethorf — COMMENTED** [review]
Looks great to me!

File: client/src/components/Navbar.js (hunk 4 above)
Comment (ChianHuei) [reply client/src/components/Navbar.js]: Oh! I will clean it.

**Comment by ChianHuei** [comment]
Now, the Notifications look better. If you click the list item, a new tab will be opened and link to the product url
Screen Shot 2020-07-16 at 4 04 10 PM

**Pull request merged**